[Edit] C++: enum (#7296)
Codecademy/docs: content/cpp/concepts/enum/enum.md

---
Title: 'enum'
Description: 'Defines a set of named integer constants in C++ for better code readability and maintainability.'
Subjects:
  - 'Code Foundations'
  - 'Computer Science'
Tags:
  - 'Data Types'
  - 'Enum'
  - 'Variables'
CatalogContent:
  - 'learn-c-plus-plus'
  - 'paths/computer-science'
---

> An **`enum`** in C++ is a user-defined [data type](https://www.codecademy.com/resources/docs/cpp/data-types) that defines a set of named integer constants. It provides a way to create symbolic names for a group of related values, making code more readable and maintainable. Enums are particularly useful when you need to represent a fixed set of options or states in your program.

Reviewer: ```suggestion
An **`enum`** in C++ is a user-defined [data type](https://www.codecademy.com/resources/docs/cpp/data-types) that defines a set of named integer constants. It provides a way to create symbolic names for a group of related values, making code more readable and maintainable. Enums are particularly useful when representing a fixed set of options or states in the program.
```

## Syntax of C++ `enum`

```pseudo
enum enum_name {
  constant1,
  constant2,
  constant3,
  ...
};
```

**Parameters:**

- `enum_name`: The name of the enumeration type
- `constant1, constant2, constant3, ...`: Named constants within the enum

**Return value:**

Enums do not return values directly. They define a new data type that can be used to create variables.

## Example 1: Basic Enum Creation in C++

> This example demonstrates how to create a simple enum and use it in a program:

Reviewer: ```suggestion
This example demonstrates how to create a simple `enum` and use it in a program:
```

  TrafficLight currentLight = RED;

  // Use enum in switch statement
  switch (currentLight) {
    case RED:
      cout << "Stop! Red light is on." << endl;
      break;
    case YELLOW:
      cout << "Caution! Yellow light is on." << endl;
      break;
    case GREEN:
      cout << "Go! Green light is on." << endl;
      break;
    default:
      cout << "Unknown light state." << endl;
  }

  // Simulate traffic light sequence
  cout << "\nTraffic light sequence:" << endl;
  for (int i = RED; i <= GREEN; i++) {
    TrafficLight light = static_cast<TrafficLight>(i);
    switch (light) {
      case RED:
        cout << "RED -> ";
        break;
      case YELLOW:
        cout << "YELLOW -> ";
        break;
      case GREEN:
        cout << "GREEN";
        break;
    }
  }
  cout << endl;

  return 0;
}
```

The output generated by this code is:

```shell
Stop! Red light is on.

Traffic light sequence:
RED -> YELLOW -> GREEN
```

This example shows how enums work perfectly with switch statements, providing clear and readable control flow. The enum values make the code self-documenting and easier to maintain than using magic numbers.

## Codebyte Example: Create Enum Class

> This example demonstrates scoped enums (enum class) introduced in C++11, which provide better type safety and scope control:

Reviewer: ```suggestion
This example demonstrates scoped enums (`enum` class) introduced in C++11, which provide better type safety and scope control:
```

```codebyte/cpp
#include <iostream>
using namespace std;

// Define enum class for better type safety
enum class Color {
  RED,
  GREEN,
  BLUE
};

enum class Size {
  SMALL,
  MEDIUM,
  LARGE
};

int main() {
  // Create enum class variables
  Color favoriteColor = Color::RED;
  Size shirtSize = Size::MEDIUM;

  // Display enum class values (requires casting)
  cout << "Favorite color code: " << static_cast<int>(favoriteColor) << endl;
  cout << "Shirt size code: " << static_cast<int>(shirtSize) << endl;

  // Enum class prevents implicit conversions
  // This would cause compilation error:
  // if (favoriteColor == Size::SMALL) { ... }

  // Use enum class in switch statement
  switch (favoriteColor) {
    case Color::RED:
      cout << "You chose red color!" << endl;
      break;
    case Color::GREEN:
      cout << "You chose green color!" << endl;
      break;
    case Color::BLUE:
      cout << "You chose blue color!" << endl;
      break;
  }

  return 0;
}
```

> **Enum class** provides several advantages over regular enums: they are strongly typed (no implicit conversion to integers), they don't pollute the surrounding namespace, and they prevent naming conflicts. You must use the scope resolution operator (`::`) to access enum class values.

Reviewer: ```suggestion
The text provides information about the strengths of the **Enum class** compared to regular enums. Key advantages include strong typing, which avoids implicit conversion to integers, reduced namespace pollution, and prevention of naming conflicts. Accessing enum class values requires the use of the scope resolution operator (`::`).
```
